lock-api-objects/src/main/java/com/palantir/lock/watch/ClientLockWatchEventLogImpl.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentSkipListMap;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.ImmutableList;
+
+public final class ClientLockWatchEventLogImpl implements ClientLockWatchEventLog {
```

Comment:
In real life, we expect there to be very little actual concurrency here (because we're batching start transactions at most like 2 or 3 threads will be in this class), but you've made it more complicated for yourself by having it. Please can you make this be a thread-unsafe class, where you access via a synchronized layer or something?